_posts/2015-03-29-organizing-data-self-encapsulation.markdown

```diff
@@ -0,0 +1,222 @@
+
+<!--break-->
+
+I'm going to start off with the first pattern in the section, which also seems to be the most lightweight: Self Encapsulate Field. I picked this one not because it's anything earth-shattering, but because code that doesn't do it drives me a little bit crazy.
+
+As an example, here's some code that surely looks familiar.
+
+{% highlight ruby %}
+class Person
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+
+  def full_name
+    "#{@first_name} #{@last_name}"
+  end
+end
+{% endhighlight %}
+
+That's an example of Direct Variable Access. It has no getters or setters, just instance variables that you access directly.
+
+More common is this, which comes from the [Ruby Style Guide](https://github.com/bbatsov/ruby-style-guide#attr_family):
+
+{% highlight ruby %}
+class Person
+  attr_reader :first_name, :last_name
+
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+end
+{% endhighlight %}
+
+I call that Mixed Variable Access. You set instance variables directly, but use getters for reading their values.
+
+After applying Self Encapsulation you wouldn't touch instance variables at all:
+
+{% highlight ruby %}
+class Person
+  attr_reader :first_name, :last_name
+
+  def initialize(first_name, last_name)
+    self.first_name = first_name
+    self.last_name = last_name
+  end
+
+  def full_name
+    "#{first_name} #{last_name}"
+  end
+
+  private
+
+  attr_writer :first_name, :last_name
+end
+{% endhighlight %}
+
+This is a minor refactoring but I bet a lot of people will have strong opinions about it. Fowler says as much, "Debates between the two [approaches] can be heated." More than one person has seen me do this and said, "Wait, what?"
+
+Self Encapsulation offers a surprising number of advantages, though. What if we introduce a sub-class of Person to represent medical doctors. With any of the examples above we could implement a Doctor's first name as:
+
+{% highlight ruby %}
+class Doctor < Person
+  def full_name
+    "Dr. #{super}"
+  end
+end
+
+Doctor.new("Marcus", "Welby").full_name
+#=> "Dr. Marcus Welby"
+{% endhighlight %}
+
+So far, so good.
+
+Then, later, you add a getter for `first_name` in Person. Maybe you have to work around some capitalization problems.
+
+{% highlight ruby %}
+class Person
+  def first_name
+    @first_name.capitalize
+  end
+end
+{% endhighlight %}
+
+Now your direct access code looks like this:
+
+{% highlight ruby %}
+class Person
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+
+  def full_name
+    "#{@first_name} #{@last_name}"
+  end
+
+  def first_name
+    @first_name.capitalize
+  end
+end
+
+class Doctor < Person
+  def full_name
+    "Dr. #{super}"
+  end
+end
+
+Doctor.new("marcus", "Welby").full_name
+#=> "Dr. Marcus Welby"
```

Comment:
This should be "Dr. marcus Welby" (note the lower case 'm').  Since #full_name is accessing the ivar directly, the capitalization done in the getter will be skipped.  Isn't that the point of this example?